vendor/k8s.io/kube-openapi/pkg/handler/handler.go

```diff
@@ -55,13 +55,38 @@ type OpenAPIService struct {
 
 	lastModified time.Time
 
-	specBytes []byte
-	specPb    []byte
-	specPbGz  []byte
+	openapiSpec *spec.Swagger
+	specBytes   []byte
+	specPb      []byte
+	specPbGz    []byte
 
 	specBytesETag string
 	specPbETag    string
 	specPbGzETag  string
+
+	jcache    jsonCache
+	pbCache   protobufCache
+	pbgzCache protobufGzipCache
+}
+
+type baseCache struct {
+	cacheClean bool
+	cacheBytes []byte
+	inputBytes []byte
+	etag       string
+}
+
+type jsonCache struct {
+	baseCache
+	spec *spec.Swagger
+}
+
+type protobufCache struct {
+	baseCache
+}
+
+type protobufGzipCache struct {
+	baseCache
```

Comment:
Rather than having all these different types, just add a "function" member in the cache object and capture the input through a closure.